clippy_lints/src/ambiguous_method_calls.rs

```diff
@@ -0,0 +1,179 @@
+    }
+
+    fn insert_call_site(&mut self, ty: Ty<'tcx>, ident: Ident, span: Span) {
+        if let Some(spans) = self.call_sites.get_mut(&(ty, ident.name)) {
+            spans.push(span);
+        } else {
+            self.call_sites.insert((ty, ident.name), vec![span]);
+        }
+    }
+}
+
+impl_lint_pass!(AmbiguousMethodCalls<'_> => [AMBIGUOUS_METHOD_CALLS]);
+
+impl<'tcx> LateLintPass<'tcx> for AmbiguousMethodCalls<'tcx> {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: FnKind<'tcx>,
+        _: &'tcx FnDecl<'_>,
+        _: &'tcx Body<'_>,
+        _: Span,
+        def_id: LocalDefId,
+    ) {
+        if let FnKind::Method(ident, _) = kind {
+            let hir_id = cx.tcx.hir().local_def_id_to_hir_id(def_id);
+            let is_trait_impl = is_trait_impl_item(cx, hir_id);
+            let parent_id = cx.tcx.hir().get_parent_item(hir_id);
+
+            // Calling type_of on a method's default impl causes an ICE
+            if let Some(hir::Node::Item(item)) = cx.tcx.hir().find(hir::HirId::from(parent_id)) {
+                if let hir::ItemKind::Trait(..) = item.kind {
+                    return;
+                }
+            }
+
+            let parent_ty = cx.tcx.type_of(parent_id.to_def_id()).skip_binder();
+            self.insert_method(is_trait_impl, parent_ty, ident);
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if let hir::ExprKind::MethodCall(path, recv, _, call_span) = &expr.kind {
+            let recv_ty = cx.typeck_results().expr_ty(recv).peel_refs();
+
+            self.insert_call_site(recv_ty, path.ident, *call_span);
+        }
+    }
+
+    fn check_crate_post(&mut self, cx: &LateContext<'tcx>) {
+        for (ty, name, span) in &self.inherent_methods {
+            let k = &(*ty, *name);
+            if self.trait_methods.contains_key(k) {
+                span_lint(
+                    cx,
+                    AMBIGUOUS_METHOD_CALLS,
+                    *self.trait_methods.get(k).unwrap(),
+                    "ambiguous trait method name",
+                );
+                span_lint_and_help(
+                    cx,
+                    AMBIGUOUS_METHOD_CALLS,
+                    *span,
+                    "ambiguous struct method name",
+                    None,
+                    "consider renaming the struct impl's method",
+                );
+
+                if let Some(spans) = self.call_sites.get(k) {
+                    for span in spans {
+                        span_lint_and_help(
+                            cx,
+                            AMBIGUOUS_METHOD_CALLS,
+                            *span,
+                            "ambiguous method call",
+                            None,
+                            "consider renaming the struct impl's method or explicitly qualifying the call site",
+                        );
+                    }
+                }
```

Comment:
This should be a single lint call. I would lint on the inherent method since that's the most likely one to rename. The trait can be trait method(s) can be called out in a note. The call sites are probably not worth mentioning due to the amount of unhelpful noise that it can generate.